lecture_notes/2017-01-24.md

```diff
@@ -0,0 +1,451 @@
+
+"Higher-Order Flow Analysis" is a better name (conveys an insight) than
+"Higher-Order Control-Flow Analysis"
+
+
+Start with data-flow analysis. It's a great idea, big hammer.
+Looking for path-invariant facts. Fact ~= theorem.
+
+> There's a flowchart on the board. Olin says he used to write BASIC
+> programs by first writing a flowchart.
+
+Example problem: available expressions. At line N, what expressions in the program
+have already been computed? (Have their value in a register, variables in the
+expression haven't been mutated.)
+
+> Olin: Every time through the loop, I've saved myself an "add". That's an optimization.
+
+
+### Program Analysis by Analogy
+
+Rice's theorem is great. "Any interesting question about Turing machines is unsolvable."
+Forget the halting problem, Rice's theorem is way better.
+
+> M: wait, what you're saying is you can't build an algorithm. But since logic
+>    is stronger than programs, you can reason in a logic about the problem.
+
+Very hard to find truths about a program.
+Suppose the space of all truths is a complex polygon. You **oculd** use logic
+to describe this space exactly, with a very big formula. Another way to describe
+the space is by enclosing it with a circle.
+
+Polygon is exact and very complicated.
+Circle is an approximation, but sound and very simple (origin, radius).
+
+> Olin:
+>   It's not like court .... not "nothing but the truth"
+
+> Olin:
+>  Another way to look at it is: how are you resolving the things you can't
+>  figure out.
+> M:
+>  question here is whether you believe models exist; if so you can ask
+>  whether "it" is true or not. If you only believe in construction, you can't
+>  say whether its true or false. Industrial things are always circles on the
+>  outside.
+
+Computable approximation is the name of the game.
+That's what math is too, finite symbolic things that talk about infinite sets.
+
+
+### Lattice of Information
+
+When you're playing the game of program analysis, you use a lattice of information.
+"More information" goes up. Each point is something "you know" about the program.
+
+Ordering is "consistent with, but weaker".
+
+Flow analyzer is a "weak method" (term from artifical intelligence).
+It's a theorem-discoverer rather than a theorem-prover.
+
+> M:
+>  In my world, you don't want ⊤ because it represents inconsistency.
+>  In program analysis, ⊤ is acceptable because that's when you approximate.
+
+Start with what "you know" about the program. Infer more, conservatively.
+Eventually, you don't learn anymore.
+
+> M:
+>  Olin is using the word "implies" in 2 senses. One is the logical "x implies y",
+>  the other is moving up this lattice in terms of the ordering. These
+>  are almost the opposite.
+
+Also beware, people tend to flip ⊤ and ⊥ on a lattice. So be careful when
+you're reading. The only solution is to go slowly and reason carefully.
+
+Anyway, all of program analysis is start from "no information" and iterate to
```

Comment:
Also say that "no information" corresponds to ⊥ and "everything" corresponds to ⊤?

Also, "iterate to a fixpoint" -> "move up the lattice and iterate to a fixpoint" ?